[SPARK-21113][CORE] Read ahead input stream to amortize disk IO cost …

sitalkedia opened the pull request:
Profiling some of our big jobs, we see that around 30% of the time is being spent in reading the spill files from disk. In order to amortize the disk IO cost, the idea is to implement a read ahead input stream which asynchronously reads ahead from the underlying input stream when specified amount of data has been read from the current buffer. It does it by maintaining two buffer - active buffer and read ahead buffer. The active buffer contains data which should be returned when a read() call is issued. The read-ahead buffer is used to asynchronously read from the underlying input stream and once the active buffer is exhausted, we flip the two buffers so that we can start reading from the read ahead buffer without being blocked in disk I/O.

## How was this patch tested?

Tested by running a job on the cluster and could see up to 8% CPU improvement.

Commit:
[SPARK-21113][CORE] Read ahead input stream to amortize disk IO cost in the Spill reader

SparkQA commented:
**[Test build #78122 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/78122/testReport)** for PR 18317 at commit [`d42000f`](https://github.com/apache/spark/commit/d42000fe288a22bbfc56b4fef8bf32273aea6218).

SparkQA commented:
**[Test build #78122 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/78122/testReport)** for PR 18317 at commit [`d42000f`](https://github.com/apache/spark/commit/d42000fe288a22bbfc56b4fef8bf32273aea6218).
 * This patch **fails RAT tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `public class ReadAheadInputStream extends InputStream `

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/78122/
Test FAILed.

Commit:
cleanup

sitalkedia commented:
cc - @rxin, @srowen, @sameeragarwal, @zsxwing,  @mridulm

SparkQA commented:
**[Test build #78124 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/78124/testReport)** for PR 18317 at commit [`f687918`](https://github.com/apache/spark/commit/f687918f8931899961c4af4c05d7c2aebbf2a84c).

Commit:
Add apache header

SparkQA commented:
**[Test build #78124 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/78124/testReport)** for PR 18317 at commit [`f687918`](https://github.com/apache/spark/commit/f687918f8931899961c4af4c05d7c2aebbf2a84c).
 * This patch **fails RAT tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/78124/
Test FAILed.

SparkQA commented:
**[Test build #78125 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/78125/testReport)** for PR 18317 at commit [`e138c0d`](https://github.com/apache/spark/commit/e138c0dfb48bdd2d4bc142ce1f620a605df14134).

sitalkedia commented:
Jenkins retest this please.

SparkQA commented:
**[Test build #78127 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/78127/testReport)** for PR 18317 at commit [`e138c0d`](https://github.com/apache/spark/commit/e138c0dfb48bdd2d4bc142ce1f620a605df14134).

Commit:
clean up on close

Commit:
fix tests

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/78131/
Test FAILed.

SparkQA commented:
**[Test build #78132 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/78132/testReport)** for PR 18317 at commit [`dbee3e1`](https://github.com/apache/spark/commit/dbee3e1d859fb72d0f09070e3a83c5bb00c3f63f).

SparkQA commented:
**[Test build #78125 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/78125/testReport)** for PR 18317 at commit [`e138c0d`](https://github.com/apache/spark/commit/e138c0dfb48bdd2d4bc142ce1f620a605df14134).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/78125/
Test FAILed.

SparkQA commented:
**[Test build #78127 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/78127/testReport)** for PR 18317 at commit [`e138c0d`](https://github.com/apache/spark/commit/e138c0dfb48bdd2d4bc142ce1f620a605df14134).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/78127/
Test FAILed.

SparkQA commented:
**[Test build #78132 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/78132/testReport)** for PR 18317 at commit [`dbee3e1`](https://github.com/apache/spark/commit/dbee3e1d859fb72d0f09070e3a83c5bb00c3f63f).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `public abstract class GenericFileInputStreamSuite `

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/78132/
Test PASSed.

jiangxb1987 commented on core/src/main/java/org/apache/spark/io/ReadAheadInputStream.java:
```diff
@@ -0,0 +1,279 @@
+      stateChangeLock.lock();
+      arr = byteBuffer.array();
+      stateChangeLock.unlock();
+      // Please note that it is safe to release the lock and read into the read ahead buffer
```
The `byteBuffer` is not necessarily the read ahead buffer, it can be the active buffer if read is called for the first time.

sitalkedia replied:
Updated the comment accordingly.

jiangxb1987 commented on core/src/main/java/org/apache/spark/io/ReadAheadInputStream.java:
```diff
@@ -0,0 +1,279 @@
+    if (offset < 0 || len < 0 || offset + len < 0 || offset + len > b.length) {
+      throw new IndexOutOfBoundsException();
+    }
+    if (!activeBuffer.hasRemaining() && !isReadInProgress) {
```
How about:
```
     if (!activeBuffer.hasRemaining()) {
     	if (!isReadInProgress) {
     		// This condition will only be triggered for the first time read is called.
       		readAsync(activeBuffer);
     	}
     	waitForAsyncReadComplete();
     }
```

sitalkedia replied:
done.

jiangxb1987 commented on core/src/main/java/org/apache/spark/util/collection/unsafe/sort/UnsafeSorterSpillReader.java:
```diff
@@ -73,7 +73,9 @@ public UnsafeSorterSpillReader(
-        new NioBufferedFileInputStream(file, (int) bufferSizeBytes);
+        new ReadAheadInputStream(
+            new NioBufferedFileInputStream(file, (int) bufferSizeBytes),
+            (int)bufferSizeBytes, (int)bufferSizeBytes / 2);
```
Do we have a test on what would be the best threshold? Should we consider make this configurable?

sitalkedia replied:
Our test shows best result was obtained when threshold = 0.5* buffer size. I made it configurable anyways in case someone wants to play with it.

jiangxb1987 reviewed (commented):
I like this idea, left a few comments.

jiangxb1987 commented:
retest this please.

SparkQA commented:
**[Test build #79334 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/79334/testReport)** for PR 18317 at commit [`dbee3e1`](https://github.com/apache/spark/commit/dbee3e1d859fb72d0f09070e3a83c5bb00c3f63f).

shaneknapp commented:
i'm going to kill the current build and restart it once i finish doing some package upgrades.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/79334/
Test FAILed.

shaneknapp commented:
test this please

SparkQA commented:
**[Test build #79336 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/79336/testReport)** for PR 18317 at commit [`dbee3e1`](https://github.com/apache/spark/commit/dbee3e1d859fb72d0f09070e3a83c5bb00c3f63f).

SparkQA commented:
**[Test build #79336 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/79336/testReport)** for PR 18317 at commit [`dbee3e1`](https://github.com/apache/spark/commit/dbee3e1d859fb72d0f09070e3a83c5bb00c3f63f).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `public abstract class GenericFileInputStreamSuite `

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/79336/
Test PASSed.

jiangxb1987 commented:
ping @sitalkedia

kiszk commented on core/src/main/java/org/apache/spark/io/ReadAheadInputStream.java:
```diff
@@ -0,0 +1,279 @@
+   */
+  private int readInternal(byte[] b, int offset, int len) throws IOException {
+
+    if (offset < 0 || len < 0 || offset + len < 0 || offset + len > b.length) {
```
It may take care an underflow of `offset + len` (e.g. offset = 0x40000000, len= 0x40000000, and b.length = 0x50000000).

sitalkedia replied:
+1

kiszk commented on core/src/main/java/org/apache/spark/io/ReadAheadInputStream.java:
```diff
@@ -0,0 +1,279 @@
+
+  private final byte[] oneByte = new byte[1];
+
+  public ReadAheadInputStream(InputStream inputStream, int bufferSizeInBytes, int readAheadThresholdInBytes) {
```
Can we add @param for these parameters?

sitalkedia replied:
done

kiszk commented on core/src/main/java/org/apache/spark/io/ReadAheadInputStream.java:
```diff
@@ -0,0 +1,279 @@
+    }
+    if (available() >= n) {
+      // we can skip from the internal buffers
+      int toSkip = (int)n;
```
Is this cast necessary?

kiszk replied:
May need to take care for underflow (e.g. n = 0x80000000) by using [Ints.checkedCast()](https://google.github.io/guava/releases/22.0/api/docs/com/google/common/primitives/Ints.html#checkedCast-long-).

kiszk replied:
@sitalkedia How about this? Is this code available to ensure `n < 0x80000000`?

sitalkedia replied:
As @mridulm pointed, since n < available() which returns an int, so n is guaranteed to be within integer range. So we should not be worried about the casting here.

kiszk replied:
Got it. I overlooked [this comment](https://github.com/apache/spark/pull/18317#discussion_r131502049).

kiszk commented on core/src/main/java/org/apache/spark/io/ReadAheadInputStream.java:
```diff
@@ -0,0 +1,279 @@
+   * the stateChangeLock is already acquired in the caller before calling this function.
+   */
+  private int readInternal(byte[] b, int offset, int len) throws IOException {
+
```
Can we insert `assert(stateChangeLock.isLocked())` after declaring `stateChangeLock` as `ReentrantLock`?

sitalkedia replied:
done

kiszk commented on core/src/main/java/org/apache/spark/io/ReadAheadInputStream.java:
```diff
@@ -0,0 +1,279 @@
+   * the stateChangeLock is already acquired in the caller before calling this function.
+   */
+  private long skipInternal(long n) throws IOException {
+    if (n <= 0L) {
```
Can we insert `assert(stateChangeLock.isLocked())` after declaring `stateChangeLock` as `ReentrantLock`?

sitalkedia replied:
done

sitalkedia commented:
@jiangxb1987 - Sorry haven't gotten time to work on this lately. Will address comments in next few days.

Commit:
Address review comments

SparkQA commented:
**[Test build #80267 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/80267/testReport)** for PR 18317 at commit [`42740be`](https://github.com/apache/spark/commit/42740be3b174c7b6978e77105084e0d8d70fdaa3).

Commit:
Explicit lock condition check

SparkQA commented:
**[Test build #80268 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/80268/testReport)** for PR 18317 at commit [`a83a3d2`](https://github.com/apache/spark/commit/a83a3d20694126a1d1b4c4b9c28e8362713b60f8).

mridulm commented on core/src/main/java/org/apache/spark/io/ReadAheadInputStream.java:
```diff
@@ -0,0 +1,289 @@
+    }
+    if (available() >= n) {
+      // we can skip from the internal buffers
+      int toSkip = Ints.checkedCast(n);
```
In general, checkedCast must be used when parameter must be an Int as a precondition [1] - not when it is being coerced into an int as part of impl detail.
It is valid for n to be > Integer.MAX_VALUE as a parameter value.

Having said that : available() is an int - so if n is less than available, it is within Int bounds.
Using Ints.checkedCast here would be incorrect from correctness and logical point of view.

[1] checkedCast will throw an exception - which would have been the wrong behavior here given contract says n is long.

SparkQA commented:
**[Test build #80267 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/80267/testReport)** for PR 18317 at commit [`42740be`](https://github.com/apache/spark/commit/42740be3b174c7b6978e77105084e0d8d70fdaa3).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/80267/
Test FAILed.

SparkQA commented:
**[Test build #80268 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/80268/testReport)** for PR 18317 at commit [`a83a3d2`](https://github.com/apache/spark/commit/a83a3d20694126a1d1b4c4b9c28e8362713b60f8).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/80268/
Test FAILed.

Commit:
Fix tests

SparkQA commented:
**[Test build #80272 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/80272/testReport)** for PR 18317 at commit [`89cbac8`](https://github.com/apache/spark/commit/89cbac877eefe0e0dc5d5956331c228be0382f72).

Commit:
Minor review comments

SparkQA commented:
**[Test build #80273 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/80273/testReport)** for PR 18317 at commit [`d84269d`](https://github.com/apache/spark/commit/d84269d7208ccacb20795d5a46fcca52584bcbef).

SparkQA commented:
**[Test build #80273 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/80273/testReport)** for PR 18317 at commit [`d84269d`](https://github.com/apache/spark/commit/d84269d7208ccacb20795d5a46fcca52584bcbef).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/80273/
Test FAILed.

SparkQA commented:
**[Test build #80272 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/80272/testReport)** for PR 18317 at commit [`89cbac8`](https://github.com/apache/spark/commit/89cbac877eefe0e0dc5d5956331c228be0382f72).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/80272/
Test PASSed.

kiszk commented on core/src/main/java/org/apache/spark/io/ReadAheadInputStream.java:
```diff
@@ -0,0 +1,288 @@
+  @Override
+  public synchronized int available() throws IOException {
+    stateChangeLock.lock();
+    int val = activeBuffer.remaining() + readAheadBuffer.remaining();
```
Can we always ensure `val >=0`? I am afraid about overflow of `+` of two ints.

kiszk replied:
@sitalkedia Is `val` always positive?

sitalkedia replied:
Good point, added a check

sitalkedia commented:
ping @jiangxb1987

jiangxb1987 commented on core/src/main/java/org/apache/spark/io/ReadAheadInputStream.java:
```diff
@@ -0,0 +1,288 @@
+   */
+  public ReadAheadInputStream(InputStream inputStream, int bufferSizeInBytes, int readAheadThresholdInBytes) {
+    Preconditions.checkArgument(bufferSizeInBytes > 0, "bufferSizeInBytes should be greater than 0");
+    Preconditions.checkArgument(readAheadThresholdInBytes > 0 && readAheadThresholdInBytes < bufferSizeInBytes,
```
nit: Do we want to ensure that the line length is below 100? @cloud-fan

jiangxb1987 reviewed (commented):
LGTM overall.

Commit:
Wrap the read ahead input stream around the compressed input stream so that we can amortize the cost of decompression as well. Some minor comments

Commit:
Fix white space

sitalkedia commented:
@jiangxb1987 - Made a change to  wrap the read ahead input stream around the compressed input stream so that we can amortize the cost of decompression as well

SparkQA commented:
**[Test build #81042 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/81042/testReport)** for PR 18317 at commit [`6e1d563`](https://github.com/apache/spark/commit/6e1d5630cd9dab140154c70357e3513fdd596b05).

Commit:
Make sure no int overflow in available

SparkQA commented:
**[Test build #81044 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/81044/testReport)** for PR 18317 at commit [`0685699`](https://github.com/apache/spark/commit/0685699cfc90a66737d29a1020c55cc563a0965d).

mridulm commented on core/src/main/java/org/apache/spark/io/ReadAheadInputStream.java:
```diff
@@ -0,0 +1,292 @@
+    // Make sure we have no integer overflow.
+    int val = (int) Math.min((long) Integer.MAX_VALUE,
+            (long) activeBuffer.remaining() + readAheadBuffer.remaining());
+    stateChangeLock.unlock();
```
Actually, looking at available() contract, 
`Returns an estimate of the number of bytes that can be read (or skipped over) from this input stream without blocking by the *next invocation of a method* for this input stream.`.

We cannot read `activeBuffer.remaining() + readAheadBuffer.remaining()` in a single invocation - only activeBuffer.remaining(), right ? We can read upto `activeBuffer.remaining() + readAheadBuffer.remaining()` (for example) in two non blocking invocations, but that is slightly different from the expectation.

This is a subtle distinction, and it might be fine to continue with what we have here.

mridulm commented on core/src/main/java/org/apache/spark/io/ReadAheadInputStream.java:
```diff
@@ -0,0 +1,292 @@
+    assert (stateChangeLock.isLocked());
+    if (n <= 0L) {
+      return 0L;
+    }
```
Move this into `skip` itself, before lock is acquired.

sitalkedia replied:
done

mridulm commented on core/src/main/java/org/apache/spark/io/ReadAheadInputStream.java:
```diff
@@ -0,0 +1,292 @@
+      while (toSkip > 0) {
+        int skippedBytes = read(temp, 0, toSkip);
+        toSkip -= skippedBytes;
+      }
```
Do we need to actually `read` (which involves data copy) ? Simply moving limit's should suffice ?
Note: the suffix invariants in `readInternal` should be handled - namely scheduling async read and buffer flipping - which could be a moved into a method in itself.

sitalkedia replied:
hmm actually that was making the code a little simpler at the expense of data copy. Changed it to avoid any data copy.

mridulm commented on core/src/main/java/org/apache/spark/io/ReadAheadInputStream.java:
```diff
@@ -0,0 +1,292 @@
+    }
+    if (isReadInProgress) {
+      waitForAsyncReadComplete();
+    }
```
This is not always required - for example:
bufferSizeInBytes = 10240
readAheadThresholdInBytes = 1024
buffer.remaining() = 512
n = 10

If this will complicate code too much, it is fine to avoid handling for now.

mridulm commented on core/src/main/java/org/apache/spark/io/ReadAheadInputStream.java:
```diff
@@ -0,0 +1,292 @@
+    activeBuffer.flip();
+    readAheadBuffer.position(0);
+    readAheadBuffer.flip();
+    long skippedFromInputStream = underlyingInputStream.skip(toSkip);
```
Trigger asyncRead here ?

sitalkedia replied:
Good point, actually after adding this code, I found out a critical bug in `waitForAsyncReadComplete` function call which I have fixed. Thanks!

mridulm replied:
I am curious what the issue was, can you elaborate please ? Thanks !

sitalkedia replied:
Earlier in `waitForAsyncReadComplete` function I was not actually checking if `readInProgress`. And this was causing some corner case where read was not in progress and we were just blocked on the signal. In readInternal function for example, we are issueing an async read call but it is possible that the async call finish before the main thread actually waits on the asyncReadComplete signal. so `waitForAsyncReadComplete` might be blocked forever without the check.

mridulm replied:
Ah, excellent point - I missed this.
Thanks for debugging it !

mridulm commented on core/src/main/java/org/apache/spark/io/ReadAheadInputStream.java:
```diff
@@ -0,0 +1,292 @@
+  public synchronized void close() throws IOException {
+    executorService.shutdown();
+    StorageUtils.dispose(activeBuffer);
+    StorageUtils.dispose(readAheadBuffer);
```
await termination of `executorService` ?
ReadAheadInputStream can continue to read after close is invoked : which will be happening on stream which might be closed and buffers which might be disposed.

Access to activeBuffer/readAheadBuffer should be done within `stateChangeLock` ? 
Also, delegate `close()` to underlyingInputStream ?

sitalkedia replied:
Good point, fixed it.

mridulm reviewed (commented):
@sitalkedia Thanks for the changes, I made another pass through the code and made some opinionated comments :-)

mridulm commented on core/src/main/java/org/apache/spark/io/ReadAheadInputStream.java:
```diff
@@ -0,0 +1,292 @@
+      }
+      isReadInProgress = false;
+      signalAsyncReadComplete();
+      stateChangeLock.unlock();
```
If readAborted = true and readException = null will cause issues, we can set it to some stock exception (to handle non-Exception's being thrown)

SparkQA commented:
**[Test build #81042 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/81042/testReport)** for PR 18317 at commit [`6e1d563`](https://github.com/apache/spark/commit/6e1d5630cd9dab140154c70357e3513fdd596b05).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/81042/
Test FAILed.

mridulm commented on core/src/main/java/org/apache/spark/io/ReadAheadInputStream.java:
```diff
@@ -0,0 +1,292 @@
+  }
+
+
+  private void readAsync(final ByteBuffer byteBuffer) throws IOException {
```
skip this, will just add to noise ... I added it because an NPE after lock is held can happen with byteBuffer = null for whatever reason.

SparkQA commented:
**[Test build #81044 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/81044/testReport)** for PR 18317 at commit [`0685699`](https://github.com/apache/spark/commit/0685699cfc90a66737d29a1020c55cc563a0965d).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/81044/
Test PASSed.

jiangxb1987 commented:
@sitalkedia Could you find some time to address @mridulm 's comments?

sitalkedia commented:
Sure, I will address @mridulm comment in next few days.

Commit:
Address review comments

Commit:
Address review comments

sitalkedia commented on core/src/main/java/org/apache/spark/io/ReadAheadInputStream.java:
```diff
@@ -0,0 +1,292 @@
+  }
+
+  @Override
+  public synchronized int read(byte[] b, int offset, int len) throws IOException {
```
@mridulm - This is synchronized because the stateChangeLock can be reliniquished by any read call in the waitForAsyncReadComplete function. so depending on stateChangeLock for synchronizing external apis are not thread-safe.

mridulm replied:
As I mention above, as long as state is read/updated only with the lock held, I dont see how release of lock in `waitForAsyncReadComplete` is causing issues.
`stateChangeLock` is a reentrant lock, so if at entry lock was held, on exit it will continue to be held; and if at entry it was not, at exit it will not be due to the try/finally acquired/release [1].
Did I miss something here ? Perhaps I misunderstood your comment ?

[1]
```
import java.util.concurrent.locks.ReentrantLock
val stateChangeLock = new ReentrantLock()
stateChangeLock.lock()
stateChangeLock.lock()
stateChangeLock.unlock()
stateChangeLock.isLocked
stateChangeLock.unlock()
stateChangeLock.isLocked
```

sitalkedia replied:
There are few reasons I am hesitant to remove the synchronized from the public apis.

1. Some variables like executorService, underlyingInputStream are not protected by the statechange lock. There might be a race condition when we try to execute executorService.execute call at the same time shutdown the executor service in close function call.
2. I am not sure what will be the behavior if multiple threads are waiting on  asyncReadComplete.await() call. Is the signaling thread going to wake only one of them? If yes, is it going to do it in the order? If it is not guaranteed to wake up the callers in order, then the read calls which are issued from different threads at different point of time are not guaranteed to read them in order. Is that acceptable?

mridulm replied:
`executorService` and `underlyingInputStream` are `final` right ? So there is no concurrent mods to them.
If you are worried about thread safety of these - `executorService` is thread safe (jdk contract).
In your code, `underlyingInputStream` must be invoked in order (if it is not, then we have a bug) : `synchronized` block is not protecting `underlyingInputStream` anyway.


For reads on `ReadAheadInputStream`: there is no gaurantee of ordering in case of parallel read's for InputStream's - did I miss some expectation that this class provides that gaurantee to its consumers ?

Note that I want to understand why we need `synchronized`, and if there is a need for it, sure we can keep it as such :-)
The only reason I see, as of now, is that use of `oneByte` array is problematic : if we eliminate that via a thread local or rewrite code to handle `read()` without using `read(byte[])`, then I dont see a current need for it. Please let me know if I am missing something.

sitalkedia replied:
@mridulm - After iterating through the code several times and thinking about it I convinced myself that we do not need the `synchronized` for the public apis.  I also changed the use the `oneByte` array to be thread local so that we do not have any race-condition there.

mridulm replied:
Thanks for clarifying @sitalkedia.
I was not sure if I was missing something, particularly for the way it is being used and the asyncRead vs normal read usecases, w.r.t synchronization requirements.

I will wait for @zsxwing to also take a look, but this is looking really good !

sitalkedia commented on core/src/main/java/org/apache/spark/io/ReadAheadInputStream.java:
```diff
@@ -0,0 +1,292 @@
+  }
+
+  @Override
+  public synchronized int read() throws IOException {
```
All public apis are synchronized because we can not depend on stateChangeLock  to synchronize the api calls because the stateChangeLock can be reliniquished by any read call in the waitForAsyncReadComplete function

mridulm replied:
I am not sure I follow - if all state change and critical sections were protected by `stateChangeLock` - then why synchronized ? It looks like unnecessary overhead : having said that, it is possible that I have missed some flow where synchronized helps, I want to understand what those are.

SparkQA commented:
**[Test build #81197 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/81197/testReport)** for PR 18317 at commit [`28e767d`](https://github.com/apache/spark/commit/28e767d0d3e06369b1d5e3bad7c83854b7467bc2).

sitalkedia commented on core/src/main/java/org/apache/spark/io/ReadAheadInputStream.java:
```diff
@@ -0,0 +1,292 @@
+
+  @GuardedBy("stateChangeLock")
+  // true if async read is in progress
+  private boolean isReadInProgress;
```
done

sitalkedia commented on core/src/main/java/org/apache/spark/io/ReadAheadInputStream.java:
```diff
@@ -0,0 +1,292 @@
+  }
+
+  private boolean isEndOfStream() {
+    if(activeBuffer.remaining() == 0 && readAheadBuffer.remaining() == 0 && endOfStream) {
```
done

sitalkedia commented on core/src/main/java/org/apache/spark/io/ReadAheadInputStream.java:
```diff
@@ -0,0 +1,292 @@
+      }
+      isReadInProgress = false;
+      signalAsyncReadComplete();
+      stateChangeLock.unlock();
```
Just curious, is it even possible that a null exception in java? AFAIK throw null will actually throw an NPE.

mridulm replied:
When I made the comment, I maent it to guard against the cases where ` readException = null` with `readAborted = true` causing NPE in case there is code like
`if (readAborted) throw readException;`
I dont think I saw some construct like that - but in case it existed, wanted to avoid it.

sitalkedia commented on core/src/main/java/org/apache/spark/io/ReadAheadInputStream.java:
```diff
@@ -0,0 +1,292 @@
+   */
+  private int readInternal(byte[] b, int offset, int len) throws IOException {
+    assert (stateChangeLock.isLocked());
+    if (offset < 0 || len < 0 || offset + len < 0 || offset + len > b.length) {
```
done

mridulm commented on core/src/main/java/org/apache/spark/io/ReadAheadInputStream.java:
```diff
@@ -0,0 +1,317 @@
+      return true;
+    }
+    return  false;
+  }
```
I meant `hasRemaining` on the buffer's instead of `remaining() == 0`.
The method name `isEndOfStream` looks more appropriate (and is consistent with the return value).

sitalkedia replied:
I see. My bad, changed accordingly

mridulm commented on core/src/main/java/org/apache/spark/io/ReadAheadInputStream.java:
```diff
@@ -0,0 +1,317 @@
+      public void run() {
+        stateChangeLock.lock();
+        byte[] arr = byteBuffer.array();
+        stateChangeLock.unlock();
```
The lock/unlock here can be removed - and `final byte[] arr` moved to earlier `stateChangeLock` locked block in readAsync.

sitalkedia replied:
done

mridulm commented on core/src/main/java/org/apache/spark/io/ReadAheadInputStream.java:
```diff
@@ -0,0 +1,317 @@
+        // So there is no race condition in both the situations.
+        boolean handled = false;
+        int read = 0;
+        Exception exception = new Exception("Unknown exception in ReadAheadInputStream");
```
We should do this only in  `! handled` block - so that we create Exception object only when required - and not for each `readAsync`.

sitalkedia replied:
good point, changed

mridulm reviewed (commented):
Given the sensitive nature of the change, can you take a look at it @zsxwing if you have the bandwidth ?
I want to make sure we are not missing any corner cases !

SparkQA commented:
**[Test build #81197 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/81197/testReport)** for PR 18317 at commit [`28e767d`](https://github.com/apache/spark/commit/28e767d0d3e06369b1d5e3bad7c83854b7467bc2).
 * This patch **fails due to an unknown error code, -9**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/81197/
Test FAILed.

Commit:
Review comments

SparkQA commented:
**[Test build #81227 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/81227/testReport)** for PR 18317 at commit [`c7b60a5`](https://github.com/apache/spark/commit/c7b60a5736189ce856f7f5256391a1c51e7ae1bd).

SparkQA commented:
**[Test build #81227 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/81227/testReport)** for PR 18317 at commit [`c7b60a5`](https://github.com/apache/spark/commit/c7b60a5736189ce856f7f5256391a1c51e7ae1bd).
 * This patch **fails SparkR unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/81227/
Test FAILed.

Commit:
Remove unneeded synchronization from public apis

SparkQA commented:
**[Test build #81238 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/81238/testReport)** for PR 18317 at commit [`8076e27`](https://github.com/apache/spark/commit/8076e271b3eb3955a3daea3dcd79c553b2b8908c).

SparkQA commented:
**[Test build #81238 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/81238/testReport)** for PR 18317 at commit [`8076e27`](https://github.com/apache/spark/commit/8076e271b3eb3955a3daea3dcd79c553b2b8908c).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/81238/
Test PASSed.

sitalkedia commented:
ping @zsxwing !

sitalkedia commented:
Hi @zsxwing, could you find some time to take a look at this PR.

zsxwing commented on core/src/main/java/org/apache/spark/io/ReadAheadInputStream.java:
```diff
@@ -0,0 +1,315 @@
+    }
+    if (!activeBuffer.hasRemaining()) {
+      ByteBuffer temp = activeBuffer;
+      activeBuffer = readAheadBuffer;
```
Looks like there is a race condition here: the next `read` will try to read from a buffer that is being read from the underlying input stream.

sitalkedia replied:
I don't see any race condition here. The readInternal function is protected by the stateChangeLock so only one thread is guaranteed to atomically flip the buffer. Am I missing something?

zsxwing reviewed (changes requested):
Sorry for the delay. Made one pass. Will take another look tomorrow.

zsxwing commented on core/src/main/java/org/apache/spark/io/ReadAheadInputStream.java:
```diff
@@ -0,0 +1,315 @@
+    try {
+      executorService.awaitTermination(10, TimeUnit.MILLISECONDS);
+    } catch (InterruptedException e) {
+    }
```
You can convert it to InterruptedIOException

Commit:
Address review comments

SparkQA commented:
**[Test build #81648 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/81648/testReport)** for PR 18317 at commit [`ed426f3`](https://github.com/apache/spark/commit/ed426f33f1d858c74f380218e520127b66f43453).

Commit:
Review comments

sitalkedia commented on core/src/main/java/org/apache/spark/util/collection/unsafe/sort/UnsafeSorterSpillReader.java:
```diff
@@ -72,10 +72,15 @@ public UnsafeSorterSpillReader(
         new NioBufferedFileInputStream(file, (int) bufferSizeBytes);
     try {
-      this.in = serializerManager.wrapStream(blockId, bs);
+      this.in = new ReadAheadInputStream(serializerManager.wrapStream(blockId, bs),
```
sure, done.

sitalkedia commented on core/src/main/java/org/apache/spark/util/collection/unsafe/sort/UnsafeSorterSpillReader.java:
```diff
@@ -72,10 +72,15 @@ public UnsafeSorterSpillReader(
       bufferSizeBytes = DEFAULT_BUFFER_SIZE_BYTES;
     }
 
+    final Double readAheadFraction =
```
done

SparkQA commented:
**[Test build #81649 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/81649/testReport)** for PR 18317 at commit [`36ae448`](https://github.com/apache/spark/commit/36ae4482d3193509c6dc06a79b8ebfe5a2b88752).

sitalkedia commented on core/src/main/java/org/apache/spark/io/ReadAheadInputStream.java:
```diff
@@ -0,0 +1,315 @@
+    underlyingInputStream.close();
+    stateChangeLock.lock();
+    try {
+      StorageUtils.dispose(activeBuffer);
```
You are right, but this is being done so as to release the memory as soon as we close the inputstream otherwise, it will be freed only when GC is kicked in and in case we create too many input stream and GC does not kick in, we will have some memory leak.

zsxwing replied:
`StorageUtils.dispose` is no-op for HeapByteBuffer

sitalkedia replied:
Good point, my bad. I was under impression that I am allocating a direct byte buffer instead. Fixed.

SparkQA commented:
**[Test build #81648 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/81648/testReport)** for PR 18317 at commit [`ed426f3`](https://github.com/apache/spark/commit/ed426f33f1d858c74f380218e520127b66f43453).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/81648/
Test FAILed.

SparkQA commented:
**[Test build #81649 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/81649/testReport)** for PR 18317 at commit [`36ae448`](https://github.com/apache/spark/commit/36ae4482d3193509c6dc06a79b8ebfe5a2b88752).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/81649/
Test PASSed.

Commit:
Minor cleanup

SparkQA commented:
**[Test build #81694 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/81694/testReport)** for PR 18317 at commit [`f30117e`](https://github.com/apache/spark/commit/f30117eaf5b8274cc19832c6a36acaf44adc7915).

zsxwing committed:
Close the underlying input stream safely

This commit includes the following changes:
- Close the underlying input stream only if it's not being read.
- Always read to `readAheadBuffer`.

SparkQA commented:
**[Test build #81694 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/81694/testReport)** for PR 18317 at commit [`f30117e`](https://github.com/apache/spark/commit/f30117eaf5b8274cc19832c6a36acaf44adc7915).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/81694/
Test PASSed.

zsxwing commented on core/src/main/java/org/apache/spark/io/ReadAheadInputStream.java:
```diff
@@ -0,0 +1,313 @@
+   */
+  private int readInternal(byte[] b, int offset, int len) throws IOException {
+    assert (stateChangeLock.isLocked());
+    if (!activeBuffer.hasRemaining()) {
```
This seems wrong when `activeBuffer.hasRemaining() == false && readAheadBuffer.hasRemaining() == true`. This may happen when `read` after  `skip(size of activeBuffer)`.

sitalkedia replied:
Hmm.. not really, because if we `skip(size of active buffer)` then we will flip the active buffer and read ahead buffer. So the active buffer will actually have remaining bytes.

zsxwing commented:
I opened https://github.com/sitalkedia/spark/pull/1 in your repo. Could you take a look at it?

sitalkedia commented:
Thanks for the change. Left few comments there.

sitalkedia committed:
Merge pull request #1 from zsxwing/pr18317

Close the underlying input stream safely

SparkQA commented:
**[Test build #81743 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/81743/testReport)** for PR 18317 at commit [`884c9d7`](https://github.com/apache/spark/commit/884c9d76f30531f3737d2eacca25d3d02cc4ceaa).

SparkQA commented:
**[Test build #81743 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/81743/testReport)** for PR 18317 at commit [`884c9d7`](https://github.com/apache/spark/commit/884c9d76f30531f3737d2eacca25d3d02cc4ceaa).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/81743/
Test PASSed.

zsxwing reviewed (approved):
LGTM. Thanks! Merging to master.

cxzl25 commented:
If we generate a large number of spill files, using ReadAhead's Task in `UnsafeExternalSorter` is easier to fail and OOM. By default, each spill file requires a 2MB in-heap buffer and creates a thread.
Perhaps ReadAhead is better and more stable for `BytesToBytesMap`.

Should we introduce a threshold configuration to control whether UnsafeExternalSorter uses ReadAhead to reduce memory usage on the executor?  @mridulm @cloud-fan 


org.apache.spark.util.collection.unsafe.sort.UnsafeExternalSorter#getIterator

https://github.com/apache/spark/blob/f423885f65d41632ffe0c20afa47d22d161630da/core/src/main/java/org/apache/spark/util/collection/unsafe/sort/UnsafeExternalSorter.java#L768-L773


https://github.com/apache/spark/blob/f423885f65d41632ffe0c20afa47d22d161630da/core/src/main/java/org/apache/spark/unsafe/map/BytesToBytesMap.java#L286-L287

```java
Caused by: java.lang.OutOfMemoryError: Java heap space
	at java.base/java.nio.HeapByteBuffer.<init>(HeapByteBuffer.java:71)
	at java.base/java.nio.ByteBuffer.allocate(ByteBuffer.java:391)
	at org.apache.spark.io.ReadAheadInputStream.<init>(ReadAheadInputStream.java:105)
	at org.apache.spark.util.collection.unsafe.sort.UnsafeSorterSpillReader.<init>(UnsafeSorterSpillReader.java:74)
	at org.apache.spark.util.collection.unsafe.sort.UnsafeSorterSpillWriter.getReader(UnsafeSorterSpillWriter.java:159)
	at org.apache.spark.util.collection.unsafe.sort.UnsafeExternalSorter.getIterator(UnsafeExternalSorter.java:733)
	at org.apache.spark.sql.execution.ExternalAppendOnlyUnsafeRowArray.generateIterator(ExternalAppendOnlyUnsafeRowArray.scala:173)
	at org.apache.spark.sql.execution.ExternalAppendOnlyUnsafeRowArray.generateIterator(ExternalAppendOnlyUnsafeRowArray.scala:177)
	at org.apache.spark.sql.execution.joins.OneSideOuterIterator.advanceBufferUntilBoundConditionSatisfied(SortMergeJoinExec.scala:1240)
	at org.apache.spark.sql.execution.joins.OneSideOuterIterator.advanceStream(SortMergeJoinExec.scala:1222)
	at org.apache.spark.sql.execution.joins.OneSideOuterIterator.advanceNext(SortMergeJoinExec.scala:1254)
```

mridulm commented:
This is a fairly old change, so things would have evolved since then; can you create a PR for this proposal @cxzl25 ? Will help us evaluate the impact.
